### PR TITLE
Student enrollment vulnerable to scripting #4006

### LINKED ISSUE #4006: Student enrollment vulnerable to scripting
Try enroll `Alice Betsy <script>alert("Hello world!");</script>`.
Also will be propagated to admin activity log page.

### PR DESCRIPTION
**Pull request by thngkaiyuan**
This PR introduces code to sanitize enrollment logs at their source (during student enrollment - instead of sanitizing all messages on the admin activity log page) because many existing log messages contain legitimate HTML code which ought to be left unsanitized in order for things to be displayed correctly.

Resolves #4006.

### CODE REVIEW

**Comment by junhaoyap** [comment]
@bamboo3250 would you be the best person to review this?

**Comment by bamboo3250** [comment]
@junhaoyap yes, I will review it :+1:

File: src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java [hunk 1]
```diff
@@ -43,7 +44,7 @@ public ActionResult execute() throws EntityDoesNotExistException {
                                                                     courseId, students, hasSection, studentsInfo);
             
             statusToAdmin = "Students Enrolled in Course <span class=\"bold\">[" 
-                            + courseId + "]:</span><br>" + (studentsInfo).replace("\n", "<br>");
+                            + courseId + "]:</span><br>" + Sanitizer.sanitizeForHtml((studentsInfo)).replace("\n", "<br>");
```
Comment (bamboo3250) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java]: In line 53 below, we have a similar statement. Please sanitize it as well. But before that, let's introduce another String variable named `sanitizedStudentsInfo` then use it for both places.

Comment (thngkaiyuan) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java]: Thanks for pointing that out. Resolved in 9902d2cd5e31056f8db8fae892b3bc097b866789.

**Comment by thyageshm** [comment]
@bamboo3250 please advise @thngkaiyuan on the first commit message format

**Comment by wkurniawan07** [comment]
@thyageshm there's nothing wrong. `Resolves #...` is also acceptable to GH for closing issue upon merge.

File: src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java [hunk 2]
```diff
@@ -28,6 +29,7 @@ public ActionResult execute() throws EntityDoesNotExistException {
         String courseId = getRequestParamValue(Const.ParamsNames.COURSE_ID);
         Assumption.assertNotNull(courseId);
         String studentsInfo = getRequestParamValue(Const.ParamsNames.STUDENTS_ENROLLMENT_INFO);
+        String sanitizedStudentsInfo = Sanitizer.sanitizeForHtml((studentsInfo));
```
Comment (bamboo3250) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java]: Could you remove redundant brackets?

Comment (thngkaiyuan) [inline src/main/java/teammates/ui/controller/InstructorCourseEnrollSaveAction.java]: Oops! My bad! Resolved in f1b37baaad1faff288fab2a915f4bc895bff1eaa.

**Comment by bamboo3250** [comment]
Looks good now. :+1:

**Comment by bamboo3250** [comment]
@thngkaiyuan Please have a look at `instructorCourseEnroll.jsp` and `instructorCourseEnrollResult.jsp`. Please apply `fn:escapeXml()` to all occurrences of `${data.enrollStudents}`. You may need to add a line `<%@ taglib uri="http://java.sun.com/jsp/jstl/functions" prefix="fn" %>` under the line `<%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>` as well.

**Comment by thngkaiyuan** [comment]
@bamboo3250: Resolved in 464cee4603708aa6da151f0570df615fa79dc2e7. Btw, I might not be seeing the big picture here but why do we need to escape XML in those input fields?

**Comment by bamboo3250** [comment]
@thngkaiyuan In `InstructorCourseEnrollSaveAction.java`, there are some lines like `InstructorCourseEnrollPageData pageData = new InstructorCourseEnrollPageData(account, courseId, studentsInfo);`
Here we use `studentsInfo` (not sanitized yet) to generate the page via those JSP files. Therefore these pages are also vulnerable to scripting if we don't sanitized properly. You can try removing those `escapeXml()` and enroll a student with this string.

`Section | Team | Name | Email | Comments 
Tut Group 1 | Team 6 | Alice Betsy "> <p style="font-size:100px">alert('Hello world!');</p>  | alice@test.com`

**Comment by thngkaiyuan** [comment]
I see, thanks @bamboo3250! :smile:

**Comment by bamboo3250** [comment]
@damithc ready for your review.

**Comment by damithc** [comment]
How about adding a test case to cover this?

**Comment by bamboo3250** [comment]
@thngkaiyuan Now you check `InstructorCourseEnrollPageUiTest.java`. We need to add a test to prevent this from happening again when someone changes the code in the future. 
Please improve the test where you find suitable (i.e. putting some scripts in both success and fail cases and verifying it displays properly). You might want to make changes to `InstructorCourseEnrollPageUiTest.json` as well.
After that please run the test. If there is any fail due to changes in HTML output, you can use Godmode to fix them. https://github.com/TEAMMATES/teammates/blob/master/devdocs/godmode.md

**Comment by thngkaiyuan** [comment]
@bamboo3250: verified that the instructor course enroll pages do not contain the inject script using the tests created in bc50abc83ee906d9ce0de695e36ceaa9bc99cde3. How should I ensure that the AdminActivityLog message does not contain the script though?

**Comment by bamboo3250** [comment]
@thngkaiyuan for now, it is not trivial to do any kind of UI testing for AdminActivityLogPage. So you can leave it there. Checking for other places should be good enough.

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 3]
```diff
@@ -941,6 +942,20 @@ public void verifyFieldValue (String fieldId, String expectedValue) {
...
+    
+    /**
+     * This method verifies that the page source does not contain the given
+     * searchString.
+     * 
+     * @param searchString
+     *            the substring that we want omitted from the page source
+     * @return the AppPage
+     */
+    public AppPage doesNotContain(String searchString) {
```
Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Maybe change to `verifyNotContain()`?

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 4]
```diff
@@ -941,6 +942,20 @@ public void verifyFieldValue (String fieldId, String expectedValue) {
         assertEquals(expectedValue,
                 browser.driver.findElement(By.id(fieldId)).getAttribute("value"));
     }
+    
+    /**
+     * This method verifies that the page source does not contain the given
```
Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: We can remove `This method`.

File: src/test/java/teammates/test/pageobjects/AppPage.java [hunk 5]
```diff
@@ -941,6 +942,20 @@ public void verifyFieldValue (String fieldId, String expectedValue) {
...
+     * searchString.
+     * 
+     * @param searchString
+     *            the substring that we want omitted from the page source
```
Comment (bamboo3250) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: `want to omit`? And let's see if we can make it inline with `@param searchString` or not.

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 3 above)
Comment (thngkaiyuan) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Resolved in 7c873bf5c6270571793df4eb6a0ccab886ab40fa.

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 4 above)
Comment (thngkaiyuan) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Resolved in f3e07bd54c2726ac5019ef1cc7a7b45288b66bd7.

File: src/test/java/teammates/test/pageobjects/AppPage.java (hunk 5 above)
Comment (thngkaiyuan) [inline src/test/java/teammates/test/pageobjects/AppPage.java]: Resolved in f3e07bd54c2726ac5019ef1cc7a7b45288b66bd7 as well.

**Pull request merged**